File: project2/World.java
import java.util.ArrayList;
import java.util.Collections;
import java.io.Serializable;

class World implements Serializable {
    private static final long serialVersionUID = 1L;
    private static final int WIDTH = 40;
    private static final int HEIGHT = 20;
    
    ArrayList<Organism> organisms = new ArrayList<>();
    ArrayList<Organism> newOrganisms = new ArrayList<>();
    Organism[][] grid = new Organism[HEIGHT][WIDTH];
    String lastCollisionMessage = "No collisions yet.";
    Human human = null;
    
    public void addOrganism(Organism org) {

        if (org.y >= 0 && org.y < HEIGHT && org.x >= 0 && org.x < WIDTH) {
            if (grid[org.y][org.x] == null) { 
                organisms.add(org);
                grid[org.y][org.x] = org;
            }
        }
    }
    
    public void queueOrganism(Organism org) {
        newOrganisms.add(org);
    }
    
    public void makeTurn() {
        Collections.sort(organisms, (a, b) -> {
            if (a.initiative != b.initiative) return b.initiative - a.initiative;
            return b.age - a.age;
        });
        
        for (Organism org : new ArrayList<>(organisms)) {
            if (org.isAlive()) org.action();
        }
        
        organisms.removeIf(o -> !o.isAlive());
        
        for (int y = 0; y < HEIGHT; y++) {
            for (int x = 0; x < WIDTH; x++) {
                grid[y][x] = null;
            }
        }
        
        for (Organism org : organisms) {
            if (org.y >= 0 && org.y < HEIGHT && org.x >= 0 && org.x < WIDTH) {
                grid[org.y][org.x] = org;
            }
        }
        
        for (Organism org : newOrganisms) {
            addOrganism(org);
        }
        newOrganisms.clear();
    }
}

File: project3/world.py
import pickle

class World:
    WIDTH = 40
    HEIGHT = 20
    
    def __init__(self):
        self.organisms = []
        self.new_organisms = []
        self.grid = [[None for _ in range(self.WIDTH)] for _ in range(self.HEIGHT)]
        self.last_collision_message = "No collisions yet."
        self.human = None
    
    def add_organism(self, org):
        if 0 <= org.y < self.HEIGHT and 0 <= org.x < self.WIDTH:
            if self.grid[org.y][org.x] is None:
                self.organisms.append(org)
                self.grid[org.y][org.x] = org
    
    def queue_organism(self, org):
        self.new_organisms.append(org)
    
    def make_turn(self):
        # Sort organisms by initiative and age
        self.organisms.sort(key=lambda x: (-x.initiative, -x.age))
        
        # Make each organism perform its action
        organisms_copy = self.organisms.copy()
        for org in organisms_copy:
            if org.is_alive():
                org.action()
        
        # Remove dead organisms
        self.organisms = [org for org in self.organisms if org.is_alive()]
        
        # Clear the grid
        self.grid = [[None for _ in range(self.WIDTH)] for _ in range(self.HEIGHT)]
        
        # Place living organisms back on the grid
        for org in self.organisms:
            if 0 <= org.y < self.HEIGHT and 0 <= org.x < self.WIDTH:
                self.grid[org.y][org.x] = org
        
        # Add new organisms
        for org in self.new_organisms:
            self.add_organism(org)
        self.new_organisms.clear()
